django/utils/encoding.py

```diff
@@ -211,7 +211,19 @@ def escape_uri_path(path):
 
 def punycode(domain):
     """Return the Punycode of the given domain if it's non-ASCII."""
-    return domain.encode("idna").decode("ascii")
+    if not domain:
+        return domain
+    try:
+        # try importing idna package for IDNA 2008 compliance
+        import idna
+
+        # enable uts46 mapping for mapping: uppercase letters, normalization, etc.
+        return idna.encode(domain, uts46=True, transitional=False).decode("ascii")
```

Comment:
The mapping involved here is a default one to ensure at least some basic transforms are applied (convert to lowercase for instance). It would be up to the application to correctly perform mapping on domain names if another is appropriate as local mapping is supported by IDNA 2008. Providing that the mapped domain would be IDNA 2008 compliant, the `uts46` flag should not have any effect.
The `transitional` flag is set to `False` because this is an UTS46 feature that was created to ensure compatibility with IDNA 2003 in the transition phase but which is not IDNA 2008 compliant.